src/MySqlConnector/MySqlClient/MySqlDataReader.cs

```diff
@@ -628,205 +165,77 @@ protected override void Dispose(bool disposing)
-					if (m_state == State.NoMoreData)
-						break;
-				}
-				else if (firstByte == 0xFB)
+				while (NextResult())
 				{
-					throw new NotSupportedException("Don't support LOCAL_INFILE_Request");
 				}
-				else
-				{
-					var reader = new ByteArrayReader(payload.ArraySegment);
-					var columnCount = (int) reader.ReadLengthEncodedInteger();
-					m_columnDefinitions = new ColumnDefinitionPayload[columnCount];
-					m_dataOffsets = new int[columnCount];
-					m_dataLengths = new int[columnCount];
-
-					for (var column = 0; column < m_columnDefinitions.Length; column++)
-					{
-						payload = await m_session.ReceiveReplyAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
-						m_columnDefinitions[column] = ColumnDefinitionPayload.Create(payload);
-					}
+				m_ResultSet = null;
 
-					payload = await m_session.ReceiveReplyAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
-					EofPayload.Create(payload);
-
-					m_command.LastInsertedId = -1;
-					m_state = State.ReadResultSetHeader;
-					break;
+				var connection = Command.Connection;
+				connection.HasActiveReader = false;
+				Command.ReaderClosed();
+				if (m_behavior.HasFlag(CommandBehavior.CloseConnection))
+				{
+					Command.Dispose();
+					connection.Close();
 				}
+				Command = null;
 			}
 		}
 
-		private void Reset()
-		{
-			m_columnDefinitions = null;
-			m_dataOffsets = null;
-			m_dataLengths = null;
-			m_state = State.None;
-		}
-
-		private void VerifyHasResult()
-		{
-			VerifyNotDisposed();
-			if (m_state != State.ReadResultSetHeader && m_state != State.ReadingRows)
-				throw new InvalidOperationException("There is no current result set.");
-		}
-
-		private void VerifyRead()
-		{
-			VerifyHasResult();
-			if (m_state != State.ReadingRows)
-				throw new InvalidOperationException("Read must be called first.");
-		}
-
 		private void VerifyNotDisposed()
 		{
-			if (m_command == null)
+			if (Command == null)
 				throw new ObjectDisposedException(GetType().Name);
 		}
 
-		private static long ReadFieldLength(ByteArrayReader reader)
+		private ResultSet GetResultSet()
 		{
-			var leadByte = reader.ReadByte();
-			switch (leadByte)
-			{
-			case 0xFB:
-				return -1;
-			case 0xFC:
-				return reader.ReadFixedLengthUInt32(2);
-			case 0xFD:
-				return reader.ReadFixedLengthUInt32(3);
-			case 0xFE:
-				return checked((long) reader.ReadFixedLengthUInt64(8));
-			default:
-				return leadByte;
-			}
+			VerifyNotDisposed();
+			if (m_ResultSet == null)
+				throw new InvalidOperationException("There is no current result set.");
+			return m_ResultSet;
 		}
 
-		private enum State
+		private ResultSet _resultSet;
+		private ResultSet m_ResultSet
 		{
-			None,
-			ReadResultSetHeader,
-			AlreadyReadFirstRow,
-			ReadingRows,
-			HasMoreData,
-			NoMoreData,
+			get { return _resultSet;  }
+			set { _resultSet = value?.Actiate() ?? new ResultSet(this); }
 		}
```

Comment:
From an API design perspective, it's somewhat confusing to have `_resultSet` field, `m_ResultSet` property and `GetResultSet` method that all do something slightly different. Also, a setter that transforms `m_ResultSet = null;` into a non-null value smells like a bit of a hack that would be confusing to understand.

Perhaps the property should be removed and its setter become an explicit `SetAndActivateResultSet` method that manipulates the field?